lib/dry/validation/schema/rule.rb

```diff
@@ -110,8 +110,8 @@ def key(predicate, args = [])
           node =
             if predicate.is_a?(::Symbol)
               [target.type, [name, [:predicate, [predicate, args]]]]
-            elsif predicate.is_a?(Types::Constrained)
-              [target.type, [name, predicate.rule.to_ast]]
+            elsif predicate.respond_to?(:rule)
```

Comment:
This obviously isn't the best solution but @solnic will take a look a reimplementing when support for custom coercions is added.